docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
-.. currentmodule:: torch.sparse
+.. currentmodule:: torch
 
 .. _sparse-docs:
 
 torch.sparse
 ============
 
-.. warning::
+Torch supports sparse tensors which provide memory-efficient storage
+and processing algorithms of tensors for which the majority of
+elements have the same value (fill value). This allows saving storage
+and processing resources when the size and sparsity levels of tensors
+are high.  Otherwise, for small-sized or low-sparsity tensors using
+strided tensors is likely the most efficient approach.
 
-    This API is in beta and may change in the near future.
+.. _sparse-fill-value-docs:
 
-Torch supports sparse tensors in COO(rdinate) format, which can
-efficiently store and process tensors for which the majority of elements
-are zeros.
+Fill value
```

Comment:
Is a fill value really the most important thing to understand about sparse? I would have thought it was the formats? 

Also, tracking all the sparse docs in this PR is a little tricky. Do you expect that sparse.rst will include all the sparse documentation?

Reply:
> Is a fill value really the most important thing to understand about sparse? I would have thought it was the formats?

I think the concept of fill value is undervalued when talking about sparse arrays in general. Historically, the fill value has been assumed to be zero because this choice makes the most sense in Linear Algebra. All existing libraries implementing sparse storage formats always assume zero as the fill value.
Arrays, in general, are extensively used in many fields of numerical analysis and I think the lack of non-zero fill value is the only reason that prevents the sparse array support to be as fully featured as the strided array support currently is in many numerical software libraries. For instance, not supporting `numpy.exp` on NumPy ndarrays would be unthinkable nowadays while none of the existing sparse libraries cannot support such elementary calculations on sparse arrays in a memory-efficient way because these do not implement support for a non-zero fill value. People may accept this as an inevitable restriction but I believe we can change this situation and turn the sparse formats as an implementation detail of storing structured array data efficiently rather than considering sparse arrays as specialized data structures that represent arrays for a limited selection of fields in scientific computing.

> Also, tracking all the sparse docs in this PR is a little tricky. Do you expect that sparse.rst will include all the sparse documentation?

The reason for tracking all the sparse docs in sparse.rst is to make clear to users what functionality in PyTorch can be used with sparse tensors. Atm, only about 10% of all PyTorch functions support sparse tensors (see https://github.com/Quansight-Labs/rfcs/blob/pearu/rfc0005/RFC0003-sparse-roadmap/SparseSupportState.md).
Notice that sparse tensor objects are `torch.Tensor` instances and users ought to able to use tensors of any layout as inputs to general-purpose functionality that PyTorch supports. Only a handful of PyTorch functions (constructors, layout converters, etc) ought to be layout specific and document the supported layouts explicitly. I think we are currently far from this state and the alternative could be to document in each function what layouts the function supports or not.

In the future, I would expect that the sparse.rst would document only the supported sparse formats and sparse format related functions. The docs for the general-purpose PyTorch functions ought to live under torch.rst.

Reply:
In the new sparse.rst, the discussion on the fill value has been minimized.